docs/RFCS/sql_batching.md

```diff
@@ -0,0 +1,499 @@
+(Get, Put, etc.), since these methods are already shared across `client.DB`, `client.Txn`, and
+`client.Batch`. Our new batch delegation object could then also implement this interface to
+allow for conditional indirection in the SQL layer.
+
+This approach gets more complicated when statements that issue multiple KV batches are
+considered. These cases are not rare, and arise for all statements that deal with foreign
+keys. This is also the case for UPSERT statements. The optimal solution here is not clear,
+but may reduce to the serial execution of these parts of the statement evaluation, or the
+decision to simply not parallelize any statement that requires more than one `client.Batch`
+execution at all.
+
+### Concurrent Batches
+
+An alternative approach looks to gain concurrency support at a layer below SQL, in KV logic.
+If the restriction that only one `client.Batch` can execute in a `client.Txn` at a time was
+lifted then SQL statement execution logic would not have to change between serially executed
+statements and concurrently executed statements. Each independent statement could run in its own
+goroutine and use it's own `client.Batch`
+
+This offers a number of benefits including that it is easier to reason about from the SQL
+layer, it eliminates the issue about statements that execute multiple batches, it is
+a generally useful improvement to the KV client API that could be beneficial elsewhere,
+and it removes the restriction of [coalesced batches](#coalesced-batches) that statements
+need to access the KV layer in lockstep. This last point means that this alternative should
+also perform better.
+
+However, it is unclear how difficult lifting this restriction would be. At the very least,
+it would require `client.Txn` to manage its state in a thread-safe manner, and would also
+require changes to `kv.TxnCoordSender`.
+
+## Error Handling
+
+Error handling is a major concern with the batching of SQL statements. Regardless of whether
+statements are batched together or not, their error semantics should behave as if they were
+executed serially. This has a few consequences that can be demonstrated using an example of
+two ordered statements S<sub>1</sub> and S<sub>2</sub> that are being run in parallel. If during
+parallel execution we see an error for S<sub>2</sub>, we must still wait for all earlier
+statements (in this case, just S<sub>1</sub>) to finish. Only then can we determine which error
+to return, which should always correspond to the "earliest" error from the ordered statements
+that are being run in parallel. Likewise, if during execution we see an error for S<sub>1</sub>,
+we should attempt to actively cancel the execution of all later statements. We must also
+be sure that no statements go into effect if any of the parallel-executing statements throw
+an error. Conveniently, the atomicity property of transactions means that we can simply abort
+the transaction and achieve this behavior.
```

Comment:
There's a classic solution to address this.
When you consider a sequence of N statements `s1, s2 ... sn` to batch, you do as follows:
1. prepare an array of error results of size N
2. indicate to your batch processor that each statement `s[i]` should write its completion status into the error array at position `i`
3. when the (concurrent) execution terminates (due to an error or not), simply scan the error array and return the first non-nil error encountered.